Fix missing index when using timestamps with index

mudge opened the pull request:
The `index` option used with `timestamps` should be passed to both `column` definitions for `created_at` and `updated_at` rather than just the first.

This was happening because `Hash#delete` is used to extract the `index` option passed to `timestamps`, thereby mutating the `options` hash in-place.

matthewd commented on activerecord/test/cases/timestamp_test.rb:
```diff
@@ -425,6 +425,17 @@ def test_all_timestamp_attributes_in_model
+      indexes = connection.indexes("flowers")
+      assert_equal ["created_at", "updated_at"], indexes.flat_map(&:columns).sort
+    else
+      warn "#{connection.class} does not respond to #indexes"
```
This `warn` doesn't sound right. Either we expect all connections to support `indexes`, and we should just assume it's there, or it's fine for a connection to not, in which case we should quietly skip. (I don't happen to know which of those is true for this particular question, but I'm sure you can find some guiding precedent.)

mudge replied:
I took this from https://github.com/rails/rails/blob/master/activerecord/test/cases/adapter_test.rb#L42 but agree that it's a little too defensive: it has now been removed.

matthewd commented on activerecord/test/schema/schema.rb:
```diff
@@ -310,6 +310,11 @@ def except(adapter_names_to_exclude)
   create_table :eyes, force: true do |t|
   end
 
+  create_table :flowers, force: true do |t|
```
Can we check this without adding a new model? Perhaps just temporarily add the timestamps to an existing table?

mudge replied:
There seem to be a few different ways to create a temporary table used in the other tests but I've now gone with directly using `ActiveRecord::Base.connection.create_table` as the model is actually unnecessary.

mudge committed:
Fix missing index when using timestamps with index

The `index` option used with `timestamps` should be passed to both
`column` definitions for `created_at` and `updated_at` rather than just
the first.

This was happening because `Hash#delete` is used to extract the `index`
option passed to `timestamps`, thereby mutating the `options` hash
in-place. Now take a copy of the `options` before deleting so that the
original is not modified.

egilburg commented on activerecord/lib/active_record/connection_adapters/abstract/schema_definitions.rb:
```diff
@@ -325,9 +325,9 @@ def column(name, type, options = {})
           raise ArgumentError, "you can't redefine the primary key column '#{name}'. To define a custom primary key, pass { id: false } to create_table."
         end
 
-        index_options = options.delete(:index)
```
Given that, performance-wise, you are implicitly duping the hash in the `expect` call later, perhaps it's clearer to do `options = options.dup` at beginning of method, and then can safely mutate them later as before.

mudge replied:
Now updated to use `options.dup` explicitly.

egilburg replied:
:+1:

mudge commented:
This branch keeps conflicting with updates to the CHANGELOG: is there any way to structure it so that it can be cleanly merged even if the CHANGELOG is updated?

rafaelfranca commented:
Backported at a0fe7b7233979958afa9d555ccbe9a5a20198e59
